[Buttons] Add color prop to buttons

dbushy727 opened the pull request:
# Button Colors

TL DR; buttons can be different colors now. Just supply the `buttonColor` prop to your button and Voila!

`<Button buttonColor={COLORS.statusRed}>Click Me!</Button>`

### Check it out here --> [Review App](https://curology-radiance-pr-219.herokuapp.com/?path=/story/buttons--button)
<br>


![large](https://user-images.githubusercontent.com/5914258/79376530-28903d80-7f0f-11ea-887e-49f8eba1bd7f.gif)


<hr>

Based on the latest design for [Dupe Patients project](https://www.figma.com/file/S4CWI1jVFMxsJZWYDUaNn0/Duplicate-Patient-Support-Experience?node-id=245%3A3103) we have to render a red button. 

This PR adds the ability to supply a color prop to any of the buttons, and it will render the buttons in that color in all the various states. Please note that it is required to use one of the colors from one of the constants files. 

The story has been updated to allow a user to pick any color they like and view it. 

Certain colors are designated alternate colors for various states, while others do not. In this PR, we have linked some colors together and derive other colors where that link does not exist.

For example:
primary -> purple80 for the RoundButton background and text respectively.
statusRed -> lighter desaturated version via tinycolor
<img width="655" alt="Screen Shot 2020-04-15 at 11 46 56 AM" src="https://user-images.githubusercontent.com/5914258/79376648-52e1fb00-7f0f-11ea-8cbe-bc8924282432.png">
<img width="297" alt="Screen Shot 2020-04-15 at 11 47 18 AM" src="https://user-images.githubusercontent.com/5914258/79376654-55445500-7f0f-11ea-91f0-9810ff9d4c28.png">
<img width="1074" alt="Screen Shot 2020-04-15 at 11 47 10 AM" src="https://user-images.githubusercontent.com/5914258/79376657-55dceb80-7f0f-11ea-89ad-34b87a5e08ee.png">

<img width="842" alt="Screen Shot 2020-04-15 at 11 51 06 AM" src="https://user-images.githubusercontent.com/5914258/79376718-6d1bd900-7f0f-11ea-843f-db2ab899327f.png">

dbushy727 commented on src/shared-components/button/constants.js:
```diff
@@ -0,0 +1,25 @@
+import { COLORS } from '../../constants';
+
+// TODO: potentially break out these pairings to the color constants file
+export const textColorsAssociatedWithColors = {
```
wasnt sure if we wanted this in our actual constants/colors file since its kinda pointed to how i was using it. Maybe there's a better way to connect colors together as related.

dbushy727 commented on src/shared-components/button/components/roundButton/style.js:
```diff
@@ -62,6 +64,54 @@ export const roundButtonLoader = disabled => css`
+ * @param  string color   the current color name of the round button (e.g purple, primary, etc.)
+ * @return string         hex string of the alternate color (e.g. #efefef)
+ */
+const determineAlternateTextColor = color => {
```
i didnt put this as a util as its kind of opinionated as to whats a derived alternate color. This works great for the round button color vs text color, but this may not be the formula we want across the board, so i left this as helper functions in the roundbutton styles specifically. Happy to break this out if we need.

dbushy727 commented:
> This looks good. I would add one small thing to the docs that just says we probably want to keep usage of these buttons to be special cases only (at least for the time being)

added a comment to the docs in 14b391ed6cf3e28466ab60847646f4c5011acb9f

`We should generally try to use the default button color when possible. Only for special cases should we need to use a different button color.`

daigof commented on src/shared-components/button/index.js:
```diff
@@ -44,6 +46,7 @@ class Button extends React.Component {
   static defaultProps = {
     disabled: false,
     buttonType: 'primary',
+    color: 'purple',
```
ok so dont hate me for this but as I pointed out above I think it would be more appropriate if we pass a COLORS constant as prop value it is safer type-wise and make more consistent with our Radiance usage (for example Icons, callout, etc we pass the hex constant value or one of COLORS constants)

dbushy727 replied:
haha no dont worry, i dont hate you 😄 i get the idea of the switch over. I made the change for now but as mentioned above, would love to talk more about this for future stuff.

dbushy727 commented on docs/button/button.md:
```diff
@@ -15,6 +15,20 @@ import { CheckmarkIcon } from 'radiance-ui/lib/icons';
     <Button disabled>Disabled Button</Button>
     <Button isFullWidth>Full Width Button</Button>
 
+    <Button color="statusRed">Primary Color</Button>
```
@daigof the problem with doing ${COLORS.statusRed} is that it gives the value, not the key. I need the key name like "statusRed" not "#abc123"

dbushy727 replied:
ok, i changed the implementation to accept the color value instead. it makes the knobs dropdown work a little weird, but otherwise everything else is fine.

I will say that it feels a bit uncommon to pass a hex argument from an enumerated list rather than key names which is typical of other ui frameworks (semantic, bootstrap, foundation, material, etc)

For the sake of this pr i will make use of the `buttonColor={COLORS.statusRed}` but would def like to talk about the pros and cons offline of this.

dbushy727 commented on docs/button/linkButton.md:
```diff
@@ -27,7 +25,7 @@ import { LinkButton } from 'radiance-ui';
     </LinkButton>
   </LinkButton.Container>
-</React.Fragment>
+</React.Fragment>;
```
fixed

dbushy727 commented:
@daigof ok i believe i addressed all your concerns. The only thing that isnt 100% exactly is the story for the RoundButton. the layout (horizontal vs vertical) is different than the rest of the button pages, so i opted to conform to the current page style rather than make it appear like the other pages.

Shemoel1 commented:
100% agree with @benkolde on special case use. Looks good to me otherwise @dbushy727
